### PR TITLE
generate systemd units in tt pack

### LINKED ISSUE #145: Generate systemd units when packing rpm/deb packages with `tt pack`
Systemd units, distributed by rpm/deb packages will let users handle the unpacked environment.

### PR DESCRIPTION
**Pull request by vr009**
With this patch:
- added systemd units generating
- added helper function for initiating files from templates to utils package

Follow up #18 
Close #145

### CODE REVIEW

File: cli/pack/systemd_dir.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+func getTTBinary(packCtx *PackCtx, envSystemPath string) string {
+	if (!packCtx.TarantoolIsSystem && !packCtx.WithoutBinaries) ||
+		packCtx.WithBinaries {
+		return filepath.Join(envSystemPath, envBinPath, "tt")
+	}
+	return "tt"
+}
+
+const (
+	appUnitContentTemplate = `[Unit]
```
Comment (psergee) [inline cli/pack/systemd_dir.go:123]: Use embedded files for the stuff like this.

File: cli/pack/systemd_dir.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+		return err
+	}
+	err = util.InstantiateFileFromTemplate(appUnitPath, appUnitTemplate, contentParams)
+	if err != nil {
+		return err
+	}
+
+	appInstUnitPathTempl := filepath.Join(systemdBaseDir,
+		"{{ .EnvName }}%.service")
+	appInstUnitPath, err := util.GetTextTemplatedStr(&appInstUnitPathTempl, fileNameCtx)
```
Comment (psergee) [inline cli/pack/systemd_dir.go:63]: Why not simple sprintf?

**Review by psergee — CHANGES_REQUESTED** [review]

File: cli/pack/deb.go [hunk 3]
```diff
@@ -25,17 +25,14 @@ const (
...
-
 	debianBinaryFileContent = "2.0\n"
 
-	defaultPackageName = "bundle.deb"
-
 	PreInstScriptName  = "preinst"
 	PostInstScriptName = "postinst"
 )
 
+var defaultEnvPrefix = filepath.Join("usr", "share", "tarantool")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/deb.go:36]: Added.

File: cli/util/util.go [hunk 4]
```diff
@@ -893,3 +893,33 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+	if strings.Contains(templateContent, "<no value>") {
+		return fmt.Errorf("The passed template cannot be filled fully. " +
+			"There are missed some fields.")
+	}
+	return nil
+}
+
+// InstantiateFileFromTemplate accepts the path to file,
+// template content and parameters for its filling.
+func InstantiateFileFromTemplate(unitPath string, unitTemplate string,
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: cli/util/util.go [hunk 5]
```diff
@@ -893,3 +893,33 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+	ctx map[string]interface{}) error {
+	appUnit, err := os.Create(unitPath)
+	if err != nil {
+		return err
+	}
+
+	unitContent, err := GetTextTemplatedStr(&unitTemplate, ctx)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go:913]: Fixed.

File: cli/util/util.go [hunk 6]
```diff
@@ -893,3 +893,33 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+	}
+
+	_, err = appUnit.WriteString(unitContent)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go:922]: Fixed.

File: cli/util/util.go [hunk 7]
```diff
@@ -893,3 +893,33 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go:924]: Fixed.

File: cli/util/util.go [hunk 8]
```diff
@@ -893,3 +893,33 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
 
 	return "", os.ErrNotExist
 }
+
+// CheckTemplateCompletion checks if there are incomplete fields in the passed template.
+func CheckTemplateCompletion(templateContent string) error {
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: cli/pack/pack.go [hunk 9]
```diff
@@ -48,4 +48,6 @@ type RpmDebCtx struct {
 	Deps []string
 	// DepsFile is a path to a file of dependencies.
 	DepsFile string
+	// SystemdUnitParamsFile is a pth to file with systemd unit parameters.
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/pack.go]: Fixed.

File: cli/pack/systemd_dir.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+	baseDirPath, pathToEnv string) error {
+	log.Infof("Initializing systemd directory.")
+
+	packageName, err := getPackageName(packCtx, opts, "", false)
+	if err != nil {
+		return err
+	}
+
+	systemdBaseDir := filepath.Join(baseDirPath, "etc", "systemd", "system")
+	err = os.MkdirAll(systemdBaseDir, dirPermissions|5)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir.go [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+	contentParams, err := getUnitParams(packCtx, pathToEnv, packageName)
+	if err != nil {
+		return err
+	}
+
+	fileNameCtx := map[string]interface{}{
+		"EnvName": packageName,
+	}
+
+	appUnitPathTempl := filepath.Join(systemdBaseDir, "{{ .EnvName }}.service")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir.go (hunk 1 above)
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	return nil
+}
+
+// getUnitParams checks if there is a passed unit params file in context and
+// returns its content. Otherwise, it returns the default params.
+func getUnitParams(packCtx *PackCtx, pathToEnv,
+	envName string) (map[string]interface{}, error) {
+	pathToEnvFile := filepath.Join(pathToEnv, "tarantool.yaml")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir.go [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+	}
+	for key := range referenceParams {
+		if _, ok := contentParams[key]; !ok {
+			contentParams[key] = referenceParams[key]
+		}
+	}
+	return contentParams, nil
+}
+
+func getTTBinary(packCtx *PackCtx, envSystemPath string) string {
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Added comments.

File: cli/pack/systemd_dir.go [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+	if (!packCtx.TarantoolIsSystem && !packCtx.WithoutBinaries) ||
+		packCtx.WithBinaries {
+		return filepath.Join(envSystemPath, envBinPath, "tt")
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir_test.go [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/tarantool/tt/cli/config"
+	"github.com/tarantool/tt/cli/pack/test_helpers"
+)
+
+func Test_initSystemdDir(t *testing.T) {
+	baseTestDir := t.TempDir()
+	defer os.RemoveAll(baseTestDir)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir_test.go]: Fixed.

File: cli/pack/systemd_dir_test.go [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+const (
+	testUnitFilePartlyDefinedParamsContent = `FdLimit: 2048`
+
+	testUnitFileFullyDefinedParamsContent = `TT: /usr/bin/tt
+ConfigPath: /path/cfg/tarantool.yaml
+EnvName: customEnvName
+FdLimit: 2048
+`
+
+	expectedContentInAppUnitTest1 = `[Unit]
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir_test.go]: Fixed.

File: cli/pack/systemd_dir.go (hunk 2 above)
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir.go [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+	fileNameCtx := map[string]interface{}{
+		"EnvName": packageName,
+	}
+
+	appUnitPathTempl := filepath.Join(systemdBaseDir, "{{ .EnvName }}.service")
+	appUnitPath, err := util.GetTextTemplatedStr(&appUnitPathTempl, fileNameCtx)
+	if err != nil {
+		return err
+	}
+	err = util.CheckTemplateCompletion(appUnitTemplate)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/systemd_dir.go [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+		return err
+	}
+
+	appInstUnitPathTempl := filepath.Join(systemdBaseDir,
+		"{{ .EnvName }}%.service")
+	appInstUnitPath, err := util.GetTextTemplatedStr(&appInstUnitPathTempl, fileNameCtx)
+	if err != nil {
+		return err
+	}
+	err = util.CheckTemplateCompletion(appUnitTemplate)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/common.go [hunk 19]
```diff
@@ -459,7 +460,10 @@ func getVersion(packCtx *PackCtx, opts *config.CliOpts, defaultVersion string) s
 	if packCtx.Version == "" {
 		// Get version from git only if we are packing an application from the current directory.
 		if opts.App.InstancesEnabled == "." {
-			packageVersion, _ = util.CheckVersionFromGit(opts.App.InstancesEnabled)
+			version, err := util.CheckVersionFromGit(opts.App.InstancesEnabled)
+			if err == nil || version != "" {
```
Comment (psergee) [inline cli/pack/common.go:464]: Is this expected that the version may have correct value and non-nil err?

File: cli/pack/deb.go [hunk 20]
```diff
@@ -25,17 +25,16 @@ const (
...
-
 	debianBinaryFileContent = "2.0\n"
 
-	defaultPackageName = "bundle.deb"
-
 	PreInstScriptName  = "preinst"
 	PostInstScriptName = "postinst"
 )
 
+// defaultEnvPrefix is a path there applications will be stored after install
```
Comment (psergee) [inline cli/pack/deb.go:34]:
```suggestion
// defaultEnvPrefix is a path where applications will be stored after install
```

File: cli/pack/systemd_dir.go [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/tarantool/tt/cli/config"
+	"github.com/tarantool/tt/cli/util"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	defaultInstanceFdLimit = 65535
+)
+
+//go:embed templates/app-unit-template.txt
```
Comment (psergee) [inline cli/pack/systemd_dir.go:19]: Please, use file extension to enable code highlight in text editor: .conf or .ini.

File: cli/pack/systemd_dir_test.go [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+	testDirs := []string{
+		filepath.Join(test1Dir, appDir),
+		filepath.Join(test2Dir, appDir),
+		filepath.Join(test3Dir, appDir),
+	}
+
+	err := test_helpers.CreateDirs(baseTestDir, testDirs)
+	require.NoError(t, err)
+
+	require.NoError(t, copy.Copy(filepath.Join("testdata", "partly-defined-params.yaml"),
```
Comment (psergee) [inline cli/pack/systemd_dir_test.go:37]: Why file are copied one-by-one? copy.Copy is able to copy fs tree.

**Comment by psergee** [comment]
Integration tests for the new functionality are missing.

File: cli/pack/systemd_dir_test.go (hunk 22 above)
Comment (vr009) [reply cli/pack/systemd_dir_test.go]: Fixed.

File: cli/pack/systemd_dir.go (hunk 21 above)
Comment (vr009) [reply cli/pack/systemd_dir.go]: Fixed.

File: cli/pack/common.go (hunk 19 above)
Comment (vr009) [reply cli/pack/common.go]: No, fixed.

File: cli/util/util.go [hunk 23]
```diff
@@ -893,3 +894,39 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
 
 	return "", os.ErrNotExist
 }
+
+// InstantiateFileFromTemplate accepts the path to file,
+// template content and parameters for its filling.
+func InstantiateFileFromTemplate(templatePath string, templateContent string,
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: cli/util/util.go [hunk 24]
```diff
@@ -893,3 +894,39 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+		removeErr := os.Remove(templatePath)
+		if removeErr != nil {
+			log.Warnf("Failed to remove a file %s", templatePath)
+		}
+		return err
+	}
+
+	parsedTemplate, err := template.New(templatePath).Parse(unitContent)
+	if err != nil {
+		return fmt.Errorf("Error parsing %s: %s", templatePath, err)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: cli/util/util.go [hunk 25]
```diff
@@ -893,3 +894,39 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+	params map[string]interface{}) error {
+	file, err := os.Create(templatePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	unitContent, err := GetTextTemplatedStr(&templateContent, params)
+	if err != nil {
+		removeErr := os.Remove(templatePath)
```
[the remark this answers is not in the document]
Comment (vr009) [reply cli/util/util.go]: Fixed.

File: doc/examples.rst [hunk 26]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
 Now the result package may be distributed and installed using dpkg command.
 The package will be installed in /usr/share/tarantool/package_name directory.
 
+RPM and Deb packaging include systemd units installation in system.
```
Comment (psergee) [inline doc/examples.rst:316]:
```suggestion
RPM and Deb packages contain systemd unit files, which are responsible for starting/stopping application instances on target system.
```

File: cli/util/util.go [hunk 27]
```diff
@@ -893,3 +893,43 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
 
 	return "", os.ErrNotExist
 }
+
+const varDefFormatError = `Wrong variable definition format: %s
+Format: var-name=value`
+
+func ParseVarDefinition(varDefText string) (struct{ Name, Value string }, error) {
```
Comment (psergee) [inline cli/util/util.go:900]: Add a comment for this public func.

File: cli/util/util.go [hunk 28]
```diff
@@ -893,3 +893,43 @@ func GetYamlFileName(fileName string, mustExist bool) (string, error) {
...
+	varDefiniton := strings.TrimSpace(strings.TrimSuffix(varDefText, "\n"))
+	varName, value, found := strings.Cut(varDefiniton, "=")
+	if !found || varName == "" || value == "" {
+		return struct{ Name, Value string }{}, fmt.Errorf(varDefFormatError, varDefText)
+	}
+	return struct{ Name, Value string }{Name: varName, Value: value}, nil
+}
+
+func ParseVarsFile(pathToFile string) (map[string]string, error) {
```
Comment (psergee) [inline cli/util/util.go:909]: Add a comment for this public func.

File: doc/examples.rst [hunk 29]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
...
+There are two units that will be installed after unpacking::
+
+   /
+   └── etc
+       └── systemd
+           └── system
+               ├── <package-name>.service
+               └── <package-name>%.service
+
+``<package-name>.service`` is responsible to start all instances of the installed environment.
```
[the remark this answers is not in the document]
Comment (vr009) [reply doc/examples.rst]: Fixed.

File: doc/examples.rst [hunk 30]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
...
+
+For starting, call:
+
+.. code-block:: bash
+
+   $ systemctl start <package-name>.service
+
+``<package-name>%.service`` is responsible to start the one instance from the environment.
```
[the remark this answers is not in the document]
Comment (vr009) [reply doc/examples.rst]: Fixed.

File: doc/examples.rst [hunk 31]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
...
+It accepts a path to a text key-value file, that looks the next:
+
+.. code-block:: text
+
+    TT=/usr/bin/tt
+    ConfigPath=/path/cfg/tarantool.yaml
+    EnvName=customEnvName
+    FdLimit=2048
+
+There is no mandatory parameters, all the missed parameters will be set to default values.
```
[the remark this answers is not in the document]
Comment (vr009) [reply doc/examples.rst]: Fixed.

File: doc/examples.rst [hunk 32]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
...
+The default values are:
+
+- TT: <tt binary, if there is no binaries in the package than its value is 'tt', otherwise a path to bin file in the host system>
```
[the remark this answers is not in the document]
Comment (vr009) [reply doc/examples.rst]: Fixed.

File: doc/examples.rst [hunk 33]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
...
+
+- ConfigPath: <the path to tarantool.yaml file in the host system>
```
[the remark this answers is not in the document]
Comment (vr009) [reply doc/examples.rst]: Target, fixed.

File: doc/examples.rst [hunk 34]
```diff
@@ -313,6 +313,53 @@ For packing deb package call:
...
+``<package-name>%.service`` is responsible to start the one instance from the environment.
+
+For starting, call:
+
+.. code-block:: bash
+
+   $ systemctl start <package-name>@<instance-name>.service
+
+For passing custom values to the unit template, use the flag ``--systemd-unit-params`` and pass a path to the file with parameters.
+It accepts a path to a text key-value file, that looks the next:
```
[the remark this answers is not in the document]
Comment (vr009) [reply doc/examples.rst]: Fixed.

File: doc/examples.rst (hunk 26 above)
Comment (vr009) [reply doc/examples.rst]: Fixed.

File: cli/util/util.go (hunk 28 above)
Comment (vr009) [reply cli/util/util.go:912]: Added.

File: cli/util/util.go (hunk 27 above)
Comment (vr009) [reply cli/util/util.go:901]: Added.

**Review by psergee — APPROVED** [review]

**Pull request merged**